_posts/2022-06-01-snitching-on-your-colleagues-using-cloud-magic.md

```diff
@@ -0,0 +1,332 @@
+
+Up for another challenge I stepped into the pretty unexplored (And not fully documented) lands of developing a task for a pipeline.
+The main objective of the task is to get the pipeline logs from a predefined pipeline run or from the current one as default.
+Those logs get filtered and send to a DynamoDB instance hosted on AWS.
+
+#### Requirements
+- Node.js
+- TFX : A packaging tool
+- Microsoft VSS Web Extension SDK package
+- Some experience writing in Javascript or Typescript
+- A Visual Studio publisher account (free)
+
+#### Getting The Logs
+This was done using the Azure Devops REST API: [Documentation](https://docs.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-7.1) 
+
+To get builds a couple of things are required:
+- The Azure Devops organization name
+- The project name
+- The build number
+- Authorization
+
+While the values for point one and two were pretty straightforward gaining access to the **build logs** required a **build id** instead of a **build number** , and since the **build id** can not be traced through the UI of Azure Devops some nested API requests were required.
+
+The Authorization was gained through the creation of a PAT token that can be used as a header in the GET Request.
+#### Filtering And Saving Failures
+Now that we have our desired logs all that remains is filtering, formatting and sending those logs to DynamoDB.
+In order to complete this operation the following steps were taken:
+- Set up authorization in a way that allows developers to use their AWS service connection.
+- Use the AWS SDK combined with the credentials from the service connection to authorize the user.
+- Filter the received logs that which had a format of plain text using regex to find possible error messages
+- Format the error messages together with the developer responsible for the build and the time of build.
+- Push our formatted object to DynamoDB
+
+#### Factory Fresh
+Eh Voilà!
+Our extension does everything we hoped it would do, so it's time implement it in our pipeline.
+
+First we compile our Typescript to Javascript since the index.js file is the default node entry point, Then we package our extension, and finally we upload it using the management portal where we share it with our organizations of choice.
+
+The result:
+
+In my publisher account:
+<img class="p-image" src="{{ '/img/2022-06-01-snitching-on-your-colleagues-using-cloud-magic/CustomTaskInMarketplace.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 100%;">
+
+In the pipeline:
+<img class="p-image" src="{{ '/img/2022-06-01-snitching-on-your-colleagues-using-cloud-magic/CustomExtensionTask.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width: 50%;">
+
```

Comment:
Can you share a repo with the code here as well? Preferably a public GitHub repo.